### PR TITLE
Add sha2 SparkSQL function

### PR DESCRIPTION
**Pull request by izchen**

### CODE REVIEW

**Comment by izchen** [comment]
friendly ping @mbasmanova. Would you mind take a look when you have some time?

File: velox/docs/functions/spark/binary.rst [hunk 1]
```diff
@@ -15,6 +15,12 @@ Binary Functions
     Computes SHA-1 digest of x and convert the result to a hex string.
     Note: x can only be varbinary type.
 
+.. spark:function:: sha2(x, bitLength) -> varchar
+
+    Calculate SHA-2 family of functions (SHA-224, SHA-256, SHA-384, and SHA-512) and convert the result to a hex string.
+    If asking for an unsupported SHA function, the return value is NULL.
```
Comment (mbasmanova) [inline velox/docs/functions/spark/binary.rst:21]: > If asking for an unsupported SHA function

What does that mean? Are you referring to unsupported values of bitLength argument? Let's explain more clearly and state which values are supported. Let's make sure to add a test case for unsupported values.

File: velox/functions/sparksql/String.h [hunk 2]
```diff
@@ -94,6 +94,59 @@ struct Sha1HexStringFunction {
...
 };
 
+/// sha2 function
+/// sha2(varbinary, bitLength) -> string
+/// Calculate SHA-2 family of functions (SHA-224, SHA-256,
+/// SHA-384, and SHA-512) and convert the result to a hex string.
+/// Returns SHA-2 digest as hex string.
+/// If asking for an unsupported SHA function, the return value is NULL.
+template <typename T>
+struct Sha2HexStringFunction {
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:104]: Why not just Sha2Function

File: velox/functions/sparksql/String.h [hunk 3]
```diff
@@ -94,6 +94,59 @@ struct Sha1HexStringFunction {
...
+    int digestLength = bitLength >> 3;
+    switch (bitLength) {
+      case 224:
+        result.resize(digestLength * 2);
+        folly::ssl::OpenSSLHash::hash(
+            folly::MutableByteRange((uint8_t*)result.data(), digestLength),
+            EVP_sha224(),
+            folly::ByteRange((const uint8_t*)input.data(), input.size()));
+        break;
+      case 0:
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:121]: This branch is confusing. Let's remove it and add logic to check for zero before the switch.

File: velox/functions/sparksql/String.h [hunk 4]
```diff
@@ -94,6 +94,59 @@ struct Sha1HexStringFunction {
...
+
+  FOLLY_ALWAYS_INLINE
+  bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varbinary>& input,
+      const int32_t& bitLength) {
+    int digestLength = bitLength >> 3;
+    switch (bitLength) {
+      case 224:
+        result.resize(digestLength * 2);
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:115]: this code is shared between case statements. Let's move it before the switch.

File: velox/functions/sparksql/String.h [hunk 5]
```diff
@@ -94,6 +94,59 @@ struct Sha1HexStringFunction {
...
+        folly::ssl::OpenSSLHash::hash(
+            folly::MutableByteRange((uint8_t*)result.data(), digestLength),
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:117]: folly::MutableByteRange((uint8_t*)result.data(), digestLength) repeats a lot; let's extract a temp variable

ditto folly::ByteRange((const uint8_t*)input.data(), input.size())

File: velox/functions/sparksql/String.h [hunk 6]
```diff
@@ -94,6 +94,59 @@ struct Sha1HexStringFunction {
...
+            folly::ByteRange((const uint8_t*)input.data(), input.size()));
+        break;
+      case 512:
+        result.resize(digestLength * 2);
+        folly::ssl::OpenSSLHash::sha512(
+            folly::MutableByteRange((uint8_t*)result.data(), digestLength),
+            folly::ByteRange((const uint8_t*)input.data(), input.size()));
+        break;
+      default:
+        return false;
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:143]: nit: perhaps, add a short comment to clarify that we return NULL if bitLength value is not supported

File: velox/functions/sparksql/tests/StringTest.cpp [hunk 7]
```diff
@@ -155,6 +162,65 @@ TEST_F(StringTest, sha1) {
       "a26704c04fc5f10db5aab58468035531cc542485");
 }
 
+TEST_F(StringTest, sha2) {
+  EXPECT_EQ(sha2(std::nullopt, 256), std::nullopt);
```
Comment (mbasmanova) [inline velox/functions/sparksql/tests/StringTest.cpp:166]: Test cases when one of the arguments is null are redundant. This is handled by the framework, not the function itself.

File: velox/functions/sparksql/tests/StringTest.cpp [hunk 8]
```diff
@@ -155,6 +162,65 @@ TEST_F(StringTest, sha1) {
...
+  EXPECT_EQ(sha2("Spark", std::nullopt), std::nullopt);
+  EXPECT_EQ(sha2(std::nullopt, std::nullopt), std::nullopt);
+  EXPECT_EQ(sha2("Spark", -999), std::nullopt);
+  EXPECT_EQ(sha2("Spark", 999), std::nullopt);
+  EXPECT_EQ(
+      sha2("", 0),
+      "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855");
```
Comment (mbasmanova) [inline velox/functions/sparksql/tests/StringTest.cpp:173]: Where do expected values come from? Would you add a comment to explain?

**Review by mbasmanova — COMMENTED** [review]
@izchen Some comments.

File: velox/functions/sparksql/String.h (hunk 2 above)
Comment (izchen) [reply velox/functions/sparksql/String.h:106]: The spark sha2 function converts the result to a base16 string, in addition to calculating the result

Comment (mbasmanova) [reply velox/functions/sparksql/String.h:106]: Thank you for clariying.

File: velox/functions/sparksql/String.h [hunk 9]
```diff
@@ -109,37 +111,27 @@ struct Sha2HexStringFunction {
       out_type<Varchar>& result,
       const arg_type<Varbinary>& input,
       const int32_t& bitLength) {
-    int digestLength = bitLength >> 3;
-    switch (bitLength) {
+    int nonzeroBitLength = (bitLength == 0) ? 256 : bitLength;
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:114]: const int32_t

File: velox/functions/sparksql/String.h [hunk 10]
```diff
@@ -109,37 +111,27 @@ struct Sha2HexStringFunction {
...
+    int digestLength = nonzeroBitLength >> 3;
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:115]: const int32_t

File: velox/functions/sparksql/String.h [hunk 11]
```diff
@@ -109,37 +111,27 @@ struct Sha2HexStringFunction {
...
+    result.resize(digestLength * 2);
+    auto hashOut =
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:117]: nit: hashOut -> resultBuffer

File: velox/functions/sparksql/String.h [hunk 12]
```diff
@@ -109,37 +111,27 @@ struct Sha2HexStringFunction {
...
+        folly::MutableByteRange((uint8_t*)result.data(), digestLength);
+    auto hashIn = folly::ByteRange((const uint8_t*)input.data(), input.size());
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:119]: nit: hashIn -> inputBuffer

File: velox/functions/sparksql/String.h [hunk 13]
```diff
@@ -109,37 +111,27 @@ struct Sha2HexStringFunction {
...
         break;
       case 512:
-        result.resize(digestLength * 2);
-        folly::ssl::OpenSSLHash::sha512(
-            folly::MutableByteRange((uint8_t*)result.data(), digestLength),
-            folly::ByteRange((const uint8_t*)input.data(), input.size()));
+        folly::ssl::OpenSSLHash::sha512(hashOut, hashIn);
         break;
       default:
+        // If asking for an unsupported bitLength, the return value is NULL.
```
Comment (mbasmanova) [inline velox/functions/sparksql/String.h:134]: nit: "If asking for" - "For"

File: velox/functions/sparksql/tests/StringTest.cpp (hunk 8 above)
Comment (izchen) [reply velox/functions/sparksql/tests/StringTest.cpp:171]: $ ./spark-3.3.1-bin-hadoop2/bin/spark-shell



scala>  def testSha2(length: Int) = println (length.toString + "\n" + spark.sql(s"select sha2('',$length), sha2('Spark',$length), sha2('0123456789abcdefghijklmnopqrstuvwxyz',$length)").collectAsList)
testSha2: (length: Int)Unit

scala>  Seq(-1, 1, 0, 224, 256, 384, 512).foreach(testSha2(_))
-1
[[null,null,null]]
1
[[null,null,null]]
0
[[e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855,529bc3b07127ecb7e53a4dcf1991d9152c24537d919178022b2c42657f79a26b,74e7e5bb9d22d6db26bf76946d40fff3ea9f0346b884fd0694920fccfad15e33]]
224
[[d14a028c2a3a2bc9476102bb288234c415a2b01f828ea62ac5b3e42f,dbeab94971678d36af2195851c0f7485775a2a7c60073d62fc04549c,e6e4a6be069cc9bead8b6050856d2b26da6b3f7efa0951e5fb3a54dd]]
256
[[e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855,529bc3b07127ecb7e53a4dcf1991d9152c24537d919178022b2c42657f79a26b,74e7e5bb9d22d6db26bf76946d40fff3ea9f0346b884fd0694920fccfad15e33]]
384
[[38b060a751ac96384cd9327eb1b1e36a21fdb71114be07434c0cc7bf63f6e1da274edebfe76f65fbd51ad2f14898b95b,1e40b8d06c248a1cc32428c22582b6219d072283078fa140d9ad297ecadf2cabefc341b857ad36226aa8d6d79f2ab67d,ce6d4ea5442bc6c830bea1942d4860db9f7b96f0e9d2c3073ffe47a0e1166d95612d840ff15e5efdd23c1f273096da32]]
512
[[cf83e1357eefb8bdf1542850d66d8007d620e4050b5715dc83f4a921d36ce9ce47d0d13c5d85f2b0ff8318d2877eec2f63b931bd47417a81a538327af927da3e,44844a586c54c9a212da1dbfe05c5f1705de1af5fda1f0d36297623249b279fd8f0ccec03f888f4fb13bf7cd83fdad58591c797f81121a23cfdd5e0897795238,95cadc34aa46b9fdef432f62fe5bad8d9f475bfbecf797d5802bb5f2937a85d93ce4857a6262b03834c01c610d74cd1215f9a466dc6ad3dd15078e3309a03a6d]]

**Comment by izchen** [comment]
@mbasmanova Thank you for the review!

**Review by mbasmanova — APPROVED** [review]

**Comment by mbasmanova** [comment]
@izchen The new test fails internally.

```
[ RUN      ] StringTest.sha2

stderr:
buck-out/dev/gen/aab7ed39/folly/range#header-mode-symlink-tree-with-header-map,headers/folly/Range.h:226:66: runtime error: addition of unsigned offset to 0x62f00000e440 overflowed to 0x62f00000e43f
    #0 0x7fe16e711edf in folly::Range<unsigned char*>::Range(unsigned char*, unsigned long) folly/Range.h:226
...
SUMMARY: UndefinedBehaviorSanitizer: pointer-overflow buck-out/dev/gen/aab7ed39/folly/range#header-mode-symlink-tree-with-header-map,headers/folly/Range.h:226:66 in 
```

**Pull request closed without merging**